agent/dns.go

```diff
@@ -509,45 +509,58 @@ RPC:
 	n := out.NodeServices.Node
 	edns := req.IsEdns0() != nil
 	addr := d.agent.TranslateAddress(datacenter, n.Address, n.TaggedAddresses)
-	records := d.formatNodeRecord(addr, req.Question[0].Name, qType, d.config.NodeTTL, edns)
+	records := d.formatNodeRecord(out.NodeServices.Node, addr, req.Question[0].Name, qType, d.config.NodeTTL, edns)
 	if records != nil {
 		resp.Answer = append(resp.Answer, records...)
 	}
 }
 
-// formatNodeRecord takes a Node and returns an A, AAAA, or CNAME record
-func (d *DNSServer) formatNodeRecord(addr, qName string, qType uint16, ttl time.Duration, edns bool) (records []dns.RR) {
+// formatTxtRecords takes a kv-map and returns it as "[k=]v" for non-empty k not starting with rfc1035-
```

Comment:
* I'd suggest that you describe what the function does and not how it does that, e.g.

> formatTxtRecords encodes the map values according to RFC 1464.

* `formatTxtRecords` isn't a good name since it doesn't return records (`[]*dns.RR`) but only the values of `TXT` records. Maybe name it `encodeKVasRFC1464` or something similar. I didn't think this all the way through but you should get the drift.

* The comment does not match the code since `k` can be empty and you'd end up with `=v`.

* RFC 1464 says that key names with special characters must be encoded:

https://www.ietf.org/rfc/rfc1464.txt
>    Attribute Names
> 
>    Any printable ASCII character is permitted for the attribute name.
>    If an equals sign is embedded in the attribute name, it must be
>    quoted with a preceding grave accent (or backquote: ``"`"``).  A
>    backquote must also be quoted with an additional ``"`"``.

also further down:

>    Leading and trailing whitespace (spaces and tabs) in the attribute
>    name are ignored unless they are quoted (with a "`").  For example,
>    "abc" matches " abc<tab>" but does not match ``"` abc"``.
> 
>    Note that most DNS server implementations require a backslash (\) or
>    double quote (") in a text string to be quoted with a preceding
>    backslash.  Accent grave (``"`"``) was chosen as a quoting character in
>    this syntax to avoid confusion with "\" (and remove the need for
>    confusing strings that include sequences like "\\\\").
> 

and

>    Attribute Values
> 
>    All printable ASCII characters are permitted in the attribute value.
>    No characters need to be quoted with a "`".  In other words, the
>    first unquoted equals sign in the TXT record is the name/value
>    delimiter.  All subsequent characters are part of the value.
> 
>    Once again, note that in most implementations the backslash character
>    is an active quoting character (and must, itself, be quoted).
> 

* I'd prefer a `switch` statement over `if/else`
* Please add a table driven test for this function especially the edge cases

* When looking at #2709 I've noticed the use of `datacenter` in the example. Would that be the consul datacenter value and if yes how does this get in here? We may want to think about some expansion language like `${DC}` or similar to fill in values from the other consul node data but that may be overkill.